refactor(themes): update themes build and output to generated

joshblack opened the pull request:
Part 2 of: https://github.com/carbon-design-system/carbon/pull/4061

This PR looks to refactor parts of our themes package to make it a little easier to work with, namely:

- Refactor `tasks/build.js` to be more file oriented. We now build files for tokens, mixins, and themes
- Move to `generated` folder approach, `scss` folder now has us defining the public API and using files from `generated`
- Add support for inlining custom properties with `custom-property` mixin
- Add support for emitting difference with `should-emit` mixin
- Refactor test setup so that tests are split per sass entrypoint file

#### Changelog

**New**

- Add support for `custom-property` and `should-emit` for part 3 of work

**Changed**

- Update public Sass API entrypoints for `themes`
- Refactor test structure for `themes`

**Removed**

joshblack committed:
feat(test-utils): add support for rgba in convert

joshblack committed:
fix(themes): remove duplicate danger token

joshblack committed:
refactor(themes): update themes build and output to generated

joshblack committed:
refactor(themes): split out tests and update import paths

joshblack committed:
chore(project): sync generated files

joshblack committed:
refactor(themes): move __tests__ out of scss dir to prevent copy

joshblack committed:
chore(project): sync generated files

joshblack commented:
Note to reviewers, spotted a regression for inline theming so will take a look at that 👀

emyarod commented on packages/components/src/globals/scss/vendor/@carbon/themes/scss/_mixins.scss:
```diff
@@ -1,170 +1,35 @@
-    --hover-field: #{map-get($theme, 'hover-field')};
+/// @access private
+/// @group @carbon/themes
+@mixin custom-property($name, $value, $prefix: $custom-property-prefix) {
```
this mixin along with `should-emit()` are also generated from `@carbon/themes` right? so we are separating these two from the `generated` directory?

joshblack replied:
@emyarod these ones are authored in `themes/scss/_mixins.scss` whereas the generated code lives under `themes/scss/generated/*` and is brought in by `themes/scss/_mixins` 👍

emyarod reviewed (commented):
one question about our generated theme files

joshblack committed:
Merge branch 'master' into refactor/update-themes-package-structure

joshblack committed:
fix(bundler): update sass check to exit with code 1

joshblack committed:
fix(themes): delay token or map initialization

We have an issue where either a token is defined first and then the theme is
updated, or the theme is defined and the token needs to be updated. This patch
updates our build logic and themes entrypoint such that this relationship holds
regardless of the import order of a token or `$carbon--theme`.

joshblack committed:
refactor(components): update test formatting

joshblack committed:
chore(project): check-in generated files

joshblack committed:
fix(cli): update sass check path to ignore fixtures

joshblack committed:
Merge branch 'master' into refactor/update-themes-package-structure

joshblack committed:
Merge branch 'master' into refactor/update-themes-package-structure

joshblack committed:
Merge branch 'master' into refactor/update-themes-package-structure

emyarod reviewed (approved):
looks good to me

mattrosno commented on packages/themes/src/tokens.js:
```diff
@@ -13,7 +13,6 @@ const colors = [
   'interactive02',
   'interactive03',
   'interactive04',
-  'danger',
```
Was `danger` token recently added upstream? Is this removal intentional?

mattrosno replied:
Oh, maybe add this to the "removed" portion of PR description.

joshblack replied:
It was added, but there was a duplicate. This removes the duplicate (still exists in the list, just further down 👍 )

mattrosno commented on packages/themes/scss/_mixins.scss:
```diff
@@ -1,170 +1,35 @@
+
+// We import mixins last so that these methods are made available to the
+// carbon--theme mixin
+@import './generated/mixins';
```
Perfect! Glad you added this so you import mixins from the themes package once and don't have to worry about if you also have to import mixins from a `/generated/mixins` file.

mattrosno commented on packages/themes/scss/_theme-maps.scss:
```diff
@@ -1,358 +1,11 @@
-$carbon--theme: $carbon--theme--white !default;
+
+// Deprecated import path
+// TODO: remove in next major release
```
Maybe go with `// TODO: remove this file in next major release (v11)` or something in case we miss this deprecation and it then hangs around longer than expected?

joshblack replied:
Down for whatever comment syntax we want to check for! No preference really

mattrosno commented on packages/themes/scss/_tokens.scss:
```diff
@@ -1,327 +1,8 @@
-// Code generated by @carbon/themes. DO NOT EDIT.
 //
```
Is it possible to rename the `_tokens.scss` files here to `_variables.scss` without a breaking change? The direction that we want to go: each package has `index.scss` and `package-name.scss` entrypoints that you choose based on your build. When using the package's Sass, imports are deterministic so if you are looking to include a Sass variable without side-effects, you `@import './path/to/variables.scss';` and you don't have to look up if the file name is actually called variables, or tokens, global-tokens, etc.

joshblack replied:
I'm definitely torn on the file naming scheme by syntax instead of feature, maybe this could be a broaden change across the project?

mattrosno replied:
It needs to be a broad change across all packages (especially components). If you're building one giant Sass file for your app, it doesn't matter because everything has previously been imported and is available. But with that, a poor development experience with slow re-builds of everything Sass.

When styling in a more modular fashion (e.g. CSS modules), each file can be its own Sass build. Great developer experience, but many more imports. "I need to use the breakpoint **mixin** for this, spacing **variables** for that, a theme **variables** here, theme **mixin** there"... deterministic imports really help.

As this PR is only about the themes package, we can defer a file naming scheme until later. My suggestion is based on what I see in the industry and that if we do it now, we don't have to do it later. It doesn't have to happen now if it requires a separate discussion. Your call.

mattrosno commented on packages/themes/scss/_mixins.scss:
```diff
@@ -1,170 +1,35 @@
-  $brand-03: map-get($theme, 'brand-03') !global;
-  $active-01: map-get($theme, 'active-01') !global;
-  $hover-field: map-get($theme, 'hover-field') !global;
+$custom-property-prefix: 'cds' !default;
```
Can this go in a `_variables.scss` file?

joshblack replied:
I wanted to keep this internal-only for now versus letting folks mess with it, and try and scope it with the mixin.

Separately, this is one reason I'm not a big fan of file naming per syntax instead of feature because functionality wise it'd be great to have these in the same file, otherwise it's hard to know where things are coming from.

mattrosno replied:
Ah, it's designated as private. I don't have a preference on where private variables live. Feel free to resolve this convo.

mattrosno commented on packages/themes/scss/_mixins.scss:
```diff
@@ -1,170 +1,35 @@
-    @include carbon--theme();
+/// @access private
+/// @group @carbon/themes
+@function should-emit($theme-a, $theme-b, $token, $emit-difference) {
```
Can this go in a `_functions.scss` file?

joshblack replied:
Same notes as above, maybe this could be a broader topic change? Thankfully we could swap any of this out in the future so shouldn't have to block anything now.

mattrosno replied:
Ah, it's designated as private. I don't have a preference on where private functions live. Feel free to resolve this convo.

mattrosno reviewed (commented):
This looks great!!! Things I didn't review in depth: build, test, bundler, cli changes.

I noticed the theme switcher add-on in this PR isn't working, React storybook. But then, for some reason I don't see it in our production storybook at all so don't know what's up there. Does that storybook add-on need updated with any of these changes?

Most of my feedback is around where the Sass lives, file names. My longest comment below goes into most detail.

emyarod commented:
I'll take a look at the theme switcher following this and #4033, but it should be working on the current `master` branch (storybook appears to be several releases behind?)

joshblack commented:
Let me dig in and see what's going on! 👀

joshblack committed:
fix(components): call carbon--theme to keep themes in sync

joshblack committed:
Merge branch 'refactor/update-themes-package-structure' of github.com:joshblack/carbon into refactor/update-themes-package-structure

joshblack commented:
@mattrosno theme switcher should be fixed 👍

joshblack committed:
chore(themes): remove comment block from theme-maps

joshblack commented:
Thanks @emyarod and @mattrosno! Will wait for netlify to finish just to verify 👀
